File: java/libs/[ITS] Data Encryption/src/net/kio/security/dataencryption/KeysGenerator.java
package net.kio.security.dataencryption;

import javax.crypto.SecretKey;
import javax.crypto.spec.SecretKeySpec;
import java.security.*;
import java.util.Base64;
import java.util.Random;

public class KeysGenerator {
    private final int asyncKeySize;
    private final int syncKeySize;
    private final String asyncAlgorithm;
    private final String syncAlgorithm;
    private KeyPairGenerator keyPairGenerator;
    private SecretKey secretKey;
    private byte[] secretKeyIv;
    private PublicKey publicKey;
    private PrivateKey privateKey;

    public KeysGenerator(int asyncKeySize, int syncKeySize, String asyncAlgorithm, String syncAlgorithm) {
        this.secretKey = null;
        this.publicKey = null;
        this.privateKey = null;
        this.asyncKeySize = asyncKeySize;
        this.syncKeySize = syncKeySize;
        this.asyncAlgorithm = asyncAlgorithm;
        this.syncAlgorithm = syncAlgorithm;
        this.initializeGenerators();
    }

    public KeysGenerator() {
        this(2048, 16);
    }

    public KeysGenerator(int asyncKeySize, int syncKeySize) {
        this(asyncKeySize, syncKeySize, "RSA/None/PKCS1PADDING", "AES/CBC/PKCS5Padding");
    }

    private void initializeGenerators() {
        try {
            this.keyPairGenerator = KeyPairGenerator.getInstance(this.asyncAlgorithm.split("/")[0]);
            this.keyPairGenerator.initialize(this.asyncKeySize);
        } catch (NoSuchAlgorithmException var2) {
            var2.printStackTrace();
        }

    }

    public void generateKeys(boolean secretKey, boolean asynchronousKeys) {
        if (secretKey) {
            this.secretKeyIv = new byte[16];
            new Random().nextBytes(this.secretKeyIv);
            byte[] secretKeyBytes = new byte[this.syncKeySize];
            new Random().nextBytes(secretKeyBytes);
            this.secretKey = new SecretKeySpec(secretKeyBytes, 0, this.syncKeySize, "AES");
        }

        if (asynchronousKeys) {
            KeyPair keyPair = this.keyPairGenerator.generateKeyPair();
            this.privateKey = keyPair.getPrivate();
            this.publicKey = keyPair.getPublic();
        }

    }

    public PrivateKey getPrivateKey() {
        return this.privateKey;
    }

    public PublicKey getPublicKey() {
        return this.publicKey;
    }

    public void setPublicKey(PublicKey publicKey) {
        this.publicKey = publicKey;
    }

    public String getStringPublicKey() {
        return Base64.getEncoder().encodeToString(publicKey.getEncoded());

    }

    public SecretKey getSecretKey() {
        return this.secretKey;
    }

    public void setSecretKey(SecretKey secretKey) {
        this.secretKey = secretKey;
    }

    public byte[] getSecretKeyIv() {
        return secretKeyIv;
    }

    public void setSecretKeyIv(byte[] secretKeyIv) {
        this.secretKeyIv = secretKeyIv;
    }

    public String getAsyncAlgorithm() {
        return asyncAlgorithm;
    }

    public String getSyncAlgorithm() {
        return syncAlgorithm;
    }

    public KeysGenerator clone() throws CloneNotSupportedException {
        return (KeysGenerator) super.clone();
    }
}

File: libs/data_encryption/keys_generator.go
package data_encryption

import (
	"crypto/rand"
	"crypto/rsa"

	"github.com/kataras/golog"
)

type KeysGenerator struct {
	asyncKeySize int
	syncKeySize  int
	SecretKey    []byte
	secretKeyIv  []byte
	PublicKey    rsa.PublicKey
	PrivateKey   rsa.PrivateKey
}

func NewGenerator() KeysGenerator {
	return KeysGenerator{
		asyncKeySize: 2048,
		syncKeySize:  16,
		SecretKey:    nil,
		secretKeyIv:  nil,
	}
}

func (keysGenerator *KeysGenerator) GenerateKeys(sync, async bool) {
	if sync {
		keysGenerator.secretKeyIv = make([]byte, 16)
		if _, err := rand.Read(keysGenerator.secretKeyIv); err != nil {
			golog.Fatal("Unable to generate secretKeyIv\n", err)
		}

		keysGenerator.SecretKey = make([]byte, keysGenerator.syncKeySize)
		if _, err := rand.Read(keysGenerator.SecretKey); err != nil {
			golog.Fatal("Unable to generate secretKey\n", err)
		}
	}
	if async {
		privateKey, err := rsa.GenerateKey(rand.Reader, keysGenerator.asyncKeySize)
		if err != nil {
			golog.Fatal("Unable to generate RSA keys\n", err)
		}
		publicKey := privateKey.PublicKey

		keysGenerator.PrivateKey = *privateKey
		keysGenerator.PublicKey = publicKey
	}
}

func (keysGenerator KeysGenerator) GetSyncKeyInfo() (info []byte) {
	info = append(info, keysGenerator.secretKeyIv...)
	info = append(info, keysGenerator.SecretKey...)
	return
}
